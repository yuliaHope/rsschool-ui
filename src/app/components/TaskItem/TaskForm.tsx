import * as React from 'react';
import { IAssignmentModel } from 'core/models';

type Props = {
    status: string;
    assignment: IAssignmentModel;
    fetchAssignmentSolution: (assignment: IAssignmentModel) => void;
};

class TaskForm extends React.Component<Props> {
    private studentRepo: React.RefObject<HTMLInputElement>;
    private studentComment: React.RefObject<HTMLTextAreaElement>;
    constructor(props: Props) {
        super(props);
        this.studentRepo = React.createRef();
        this.studentComment = React.createRef();
    }

    SendStudentTask(assignment: IAssignmentModel) {
        const { fetchAssignmentSolution } = this.props;
        const studentRepoValue: string = this.studentRepo.current!.value.replace(/\s/g, '');
        if (studentRepoValue) {
            fetchAssignmentSolution(assignment);
        } else {
            alert('wrong repo');
            this.studentRepo.current!.value = '';
        }
    }

    render() {
        const { status } = this.props;
        if (status !== 'MissedDeadline' && status !== 'Checked' && status !== 'ReadyForReview') {
            const { assignment } = this.props;
            return (
                <small className="text-muted">
                    <form
                        onSubmit={ev => {
                            ev.preventDefault();
                            this.SendStudentTask(assignment);
                        }}
                    >
                        <div className="form-group">
                            <label>Choose repo</label>
                            <input
                                className="form-control form-control-sm"
                                type="text"
                                ref={this.studentRepo}
                                placeholder="Enter link"
                                pattern="https?://github.com/.+"
                            />
                        </div>
                        <div className="form-group">
                            <label>Comments</label>
                            <textarea
                                className="form-control form-control-sm"
                                ref={this.studentComment}
                                placeholder="Write comments here"
                            />
                        </div>
                        <button type="submit" className="btn btn-primary btn-sm">
                            Submit
                        </button>
                    </form>
                </small>
            );
        } else {
            return <React.Fragment />;
        }
    }
}

export default TaskForm;
