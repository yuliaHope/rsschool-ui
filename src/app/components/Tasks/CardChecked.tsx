import * as React from 'react';
import { classNames } from 'core/styles';
import { Card, CardHeader, CardBody, CardTitle, CardText } from 'reactstrap';

const cn = classNames(require('./index.scss'));

type Props = {
    score: number;
    title: string;
    description: string;
};

const CardChecked = (props: Props) => {
    const { score, title, description } = props;
    return (
        <Card className={`${cn('card')} ${score >= 100 ? cn('border-success') : cn('border-warning')}`}>
            <CardHeader className={`${cn('text-white')} ${score >= 100 ? cn('bg-success') : cn('bg-warning')}`}>
                {score >= 100 ? 'Done!' : 'Not submitted yet!'} <span className={cn('score')}>{score}%</span>
            </CardHeader>
            <CardBody className={cn('card-body')}>
                <CardTitle className={cn('card-title')}>{title}</CardTitle>
                <CardText>{description}</CardText>
            </CardBody>
        </Card>
    );
};

export default CardChecked;
