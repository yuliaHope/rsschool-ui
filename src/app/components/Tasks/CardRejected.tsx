import * as React from 'react';
import { classNames } from 'core/styles';
import { Card, CardHeader, CardBody, CardTitle, CardText } from 'reactstrap';

const cn = classNames(require('./index.scss'));

type Props = {
    title: string;
    description: string;
};

const CardRejected = (props: Props) => {
    const { title, description } = props;
    return (
        <Card className={`${cn('card')} ${cn('border-danger')}`}>
            <CardHeader className={cn('text-white bg-danger')}>The deadline has passed</CardHeader>
            <CardBody className={cn('card-body')}>
                <CardTitle className={cn('card-title')}>{title}</CardTitle>
                <CardText>{description}</CardText>
            </CardBody>
        </Card>
    );
};

export default CardRejected;
