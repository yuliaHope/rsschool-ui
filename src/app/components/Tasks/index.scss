.card,
.card-empty {
    width: 20vw;
    margin: 50px auto;
}

.card-title,
.card-empty-title {
    text-transform: uppercase;
}

.card-body {
    padding-bottom: 100px;
}

.score {
    float: right;
}

.submit-btn {
    background-color: #000;
    text-transform: uppercase;
    font-size: 14px;
    color: #fff;
}