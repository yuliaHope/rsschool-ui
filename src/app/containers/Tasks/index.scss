.tasks {
    padding: 20px;
}