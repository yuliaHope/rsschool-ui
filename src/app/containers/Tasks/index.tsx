import { fetchAssignments, submitSolution } from 'core/actions';
import { IAssignment } from 'core/models';
import { RootState } from 'core/reducers';
import { classNames } from 'core/styles';
import * as React from 'react';
import { connect } from 'react-redux';
import TaskForm from 'components/TaskForm';
import { CardDeck } from 'reactstrap';

const cn = classNames(require('./index.scss'));

const mapStateToProps = (state: RootState, props: any): AssigmentContainerProps => {
    if (state.assignments == null) {
        return props;
    }
    return {
        ...props,
        isLoading: state.assignments.isLoading,
        courseId: props.match.params.id,
        studentId: state.user.username,
        assignments: state.assignments,
    };
};

const mapDispatchToProps = (dispatch: any, props: any): AssigmentContainerProps => {
    return {
        ...props,
        onLoad: (courseId: string) => {
            dispatch(fetchAssignments(courseId));
        },
        submitTask: (assignment: IAssignment) => {
            dispatch(submitSolution(assignment));
        },
    };
};

type AssigmentContainerProps = {
    onLoad: (courseId: string) => void;
    submitTask: (assignment: IAssignment) => void;
    studentId: string;
    courseId: string;
    isLoading: boolean;
    error: boolean | undefined;
    assignments: any;
};

const Temp = {
    github: 'fdsfsd',
    score: 100,
};

class Tasks extends React.Component<AssigmentContainerProps> {
    constructor(props: AssigmentContainerProps) {
        super(props);
    }

    async componentDidMount() {
        const { courseId } = this.props;
        await this.props.onLoad(courseId);
    }

    generateTasks() {
        const { assignments } = this.props.assignments;
        const { submitTask } = this.props;
        if (assignments && assignments.length > 0) {
            return assignments.map((item: any, i: number) => {
                const props = {
                    title: item.title,
                    urlToDescription: item.urlToDescription,
                    taskId: item.taskId,
                    studentId: item.studentId,
                    status: item.status,
                    score: item.score,
                    courseId: item.courseId,
                    submit: submitTask,
                };
                return <TaskForm key={i} {...props} />;
            });
        } else {
            return 'There is no one task';
        }
    }

    render() {
        const { isLoading, error } = this.props;
        return (
            <div className={cn('tasks')}>
                <h2>Tasks</h2>
                {isLoading ? (
                    <h3>Loading...</h3>
                ) : (
                    <section>
                        <div className="row">
                            <div className="col-6">
                                <p>
                                    Your github private repository{' '}
                                    <a className="badge badge-dark" href={Temp.github}>
                                        here
                                    </a>
                                </p>
                            </div>
                            <div className="col-6 text-right">
                                <p>You are in the TOP 50 students!</p>
                                <p>Full Score: {Temp.score}</p>
                            </div>
                        </div>
                        <CardDeck>
                            {(() => {
                                if (!isLoading && !error) {
                                    return this.generateTasks();
                                }
                            })()}
                        </CardDeck>
                    </section>
                )}
            </div>
        );
    }
}

export default connect(
    mapStateToProps,
    mapDispatchToProps,
)(Tasks);
