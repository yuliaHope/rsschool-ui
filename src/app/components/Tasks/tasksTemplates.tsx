import { IAssignment, IAssignmentStatus } from 'core/models';
import * as React from 'react';
import CardChecked from './CardChecked';
import CardRejected from './CardRejected';

export enum FeedActions {
    'Assignment' = 'Assignment',
}

export const tasksTemplates: any = {
    [FeedActions.Assignment]: {
        [IAssignmentStatus.Checked]: (assignment: IAssignment): any => {
            return (
                <CardChecked
                    score={assignment.score}
                    title={assignment.taskId}
                    description={assignment.assignmentRepo}
                />
            );
        },
        [IAssignmentStatus.Rejected]: (assignment: IAssignment): any => {
            return <CardRejected title={assignment.taskId} description={assignment.assignmentRepo} />;
        },
    },
};
