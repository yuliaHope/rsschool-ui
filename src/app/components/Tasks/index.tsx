import * as React from 'react';
import { IAssignment } from 'core/models';
import { tasksTemplates } from './tasksTemplates';
import { classNames } from 'core/styles';
const cn = classNames(require('./index.scss'));
import {
    Card,
    CardHeader,
    CardBody,
    CardFooter,
    CardTitle,
    CardText,
    Form,
    FormGroup,
    Label,
    Input,
    Button,
} from 'reactstrap';

type Props = {
    assign: () => void;
    data: IAssignment[];
};

class Tasks extends React.PureComponent<Props> {
    hasData() {
        return Array.isArray(this.props.data) && this.props.data.length > 0;
    }

    render() {
        return (
            <div>
                <h2>Tasks</h2>
                <div className="row">
                    <div className="col-6">
                        <p>
                            Your github private repository{' '}
                            <a className="badge badge-dark" href="#">
                                here
                            </a>
                        </p>
                    </div>
                    <div className="col-6 text-right">
                        <p>You are in the TOP X students!</p>
                        <p>Full Score: X</p>
                    </div>
                </div>

                {this.hasData() ? (
                    this.props.data.map((assignment, i) => {
                        const template = tasksTemplates[assignment.taskId][assignment.status];
                        return (
                            <div key={i}>
                                <div>
                                    {template ? (
                                        template(assignment)
                                    ) : (
                                        <Card className={cn('card-empty')}>
                                            <CardHeader>Not submitted yet!</CardHeader>
                                            <CardBody>
                                                <CardTitle className={cn('card-empty-title')}>
                                                    {this.props.data[i].taskId}
                                                </CardTitle>
                                                <CardText>{this.props.data[i].assignmentRepo}</CardText>
                                            </CardBody>
                                            <CardFooter>
                                                <small className={cn('text-muted')}>
                                                    <Form>
                                                        <FormGroup>
                                                            <Label for="taskUrl">Choose repo</Label>
                                                            <Input
                                                                bsSize="sm"
                                                                id="taskUrl"
                                                                name="url"
                                                                placeholder="Enter link"
                                                                required={true}
                                                                type="url"
                                                            />
                                                        </FormGroup>
                                                        <FormGroup>
                                                            <Label for="taskText">Comments</Label>
                                                            <Input
                                                                bsSize="sm"
                                                                id="taskText"
                                                                name="text"
                                                                placeholder="Write comments here"
                                                                rows="5"
                                                                type="textarea"
                                                            />
                                                        </FormGroup>
                                                        <Button
                                                            onClick={this.props.assign}
                                                            className={cn('submit-btn')}
                                                        >
                                                            Submit
                                                        </Button>
                                                    </Form>
                                                </small>
                                            </CardFooter>
                                        </Card>
                                    )}
                                </div>
                            </div>
                        );
                    })
                ) : (
                    <div>Nothing to show</div>
                )}
            </div>
        );
    }
}

export default Tasks;
