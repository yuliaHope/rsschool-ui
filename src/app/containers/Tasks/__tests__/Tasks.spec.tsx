import { shallow } from 'enzyme';
import * as React from 'react';
import Tasks from '../index';
import TaskForm from 'components/TaskForm';
const { __stateGetter } = require('react-redux');

type TasksType = any;
const mockAssignments = {
    assignments: [
        {
            title: 'Test task',
            urlToDescription: 'Test description',
            taskId: 0,
            studentId: 'nikkaktus24',
            status: 'Assigned',
            score: 0,
            courseId: 'rs-course-2018q3',
        },
    ],
};

const resultValue = <TaskForm key="0" submit={jest.fn()} {...mockAssignments.assignments[0]} />;

describe('Tasks', () => {
    beforeEach(() => {
        __stateGetter.mockImplementation(() => ({}));
    });

    it('test generateTasks function ', () => {
        const output = shallow(
            <Tasks
                submitTask={jest.fn()}
                studentId={'nikkaktus24'}
                courseId={'rs-course-2018q3'}
                isLoading={false}
                error={undefined}
                assignments={mockAssignments}
            />,
        );
        const instance = output.dive().instance() as TasksType;
        const result = instance.generateTasks();
        expect(JSON.stringify(result[0])).toEqual(JSON.stringify(resultValue));
    });

    it('renders correctly if no data', () => {
        const output = shallow(<Tasks studentId={''} courseId={''} isLoading={false} error={undefined} />);
        expect(output).toMatchSnapshot();
    });
});
