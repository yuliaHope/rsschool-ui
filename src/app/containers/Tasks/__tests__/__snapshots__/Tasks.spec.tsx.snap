// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tasks renders correctly if no data 1`] = `
<Tasks
  courseId=""
  isLoading={false}
  studentId=""
/>
`;
